File: Libraries/DateTime/native/lang-csharp/source/DateTimeHelper.cs
using System;
using System.Collections.Generic;
using Interpreter.Structs;

namespace Interpreter.Libraries.DateTime
{
    internal static class DateTimeHelper
    {
        private static TimeZoneInfo GetTimeZone(object nullableTimeZoneAsObj)
        {
            if (nullableTimeZoneAsObj == null)
            {
                return TimeZoneInfo.Utc;
            }
            return (TimeZoneInfo)nullableTimeZoneAsObj;
        }

        private static readonly System.DateTime epoch = new System.DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        public static bool UnixToStructured(int[] intOut, object nullableTimeZone, double unixTime)
        {
            int unixTimeInt = (int)unixTime;
            int micros = (int)((unixTime - unixTimeInt) * 1000000);
            TimeZoneInfo tz = GetTimeZone(nullableTimeZone);
            System.DateTime utcDateTime = epoch.AddSeconds(unixTimeInt);
            TimeSpan utcOffset = tz.GetUtcOffset(utcDateTime);
            System.DateTime dt = utcDateTime.Add(utcOffset);

            intOut[0] = dt.Year;
            intOut[1] = dt.Month;
            intOut[2] = dt.Day;
            intOut[3] = dt.Hour;
            intOut[4] = dt.Minute;
            intOut[5] = dt.Second;
            intOut[6] = micros / 1000;
            intOut[7] = micros;
            intOut[8] = 1 + (int)dt.DayOfWeek;

            return true;
        }

        public static void ParseDate(int[] intOut, object nullableTimeZone, int year, int month, int day, int hour, int minute, int microseconds)
        {
            intOut[0] = 1;
            int seconds = (int)(microseconds / 1000000);
            int micros = microseconds - seconds * 1000000;
            TimeZoneInfo tz = GetTimeZone(nullableTimeZone);
            System.DateTime dt = new System.DateTime(year, month, day, hour, minute, seconds, 0, DateTimeKind.Unspecified);
            dt = TimeZoneInfo.ConvertTimeToUtc(dt, tz);
            double unixTime = dt.Subtract(epoch).TotalSeconds;
            intOut[1] = (int)unixTime;
            intOut[2] = micros;
        }

        public static bool IsDstOccurringAt(object nativeTimeZone, int unixtime)
        {
            throw new NotImplementedException();
        }

        public static object[] InitializeTimeZoneList()
        {
            throw new NotImplementedException();
        }

        public static int GetUtcOffsetAt(object nativeTimeZone, int unixTime)
        {
            TimeZoneInfo tz = GetTimeZone(nativeTimeZone);
            System.DateTime dt = new System.DateTime(1L * unixTime * TimeSpan.TicksPerSecond);
            TimeSpan offset = tz.GetUtcOffset(dt);
            int offsetSeconds = (int)offset.TotalSeconds;
            return offsetSeconds;
        }

        public static object GetDataForLocalTimeZone(string[] strOut, int[] intOut)
        {
            TimeZoneInfo tzInfo = TimeZoneInfo.Local;
            strOut[0] = tzInfo.DisplayName.Split(')')[1].Trim();
            strOut[1] = tzInfo.Id;
            intOut[0] = -1; // Not used. Remove this.
            intOut[1] = tzInfo.SupportsDaylightSavingTime ? 1 : 0;
            return tzInfo;
        }
    }
}

File: Libraries/DateTime/native/lang-java/source/DateTimeHelper.java
package org.crayonlang.libraries.datetime;

import org.crayonlang.interpreter.structs.*;

public class DateTimeHelper {
	
	private static java.util.TimeZone utcTimeZone = null;
	private static java.util.TimeZone getTimeZone(Object nullableTimeZone) {
		if (nullableTimeZone == null) {
			if (utcTimeZone == null) {
				utcTimeZone = java.util.TimeZone.getTimeZone("UTC");
			}
			return utcTimeZone;
		}
		return (java.util.TimeZone) nullableTimeZone;
	}
	
	public static Object getDataForLocalTimeZone(String[] strOut, int[] intOut) {
		java.util.TimeZone tz = java.util.Calendar.getInstance().getTimeZone();
		String name = tz.getDisplayName();
		strOut[0] = name;
		strOut[1] = tz.toZoneId().getId();
		intOut[0] = -1;
		intOut[1] = tz.observesDaylightTime() ? 1 : 0;
		return tz;
	}
	
	public static int getOffsetFromUtcNow(Object nativeTz) {
		throw new RuntimeException("Not implemented.");
	}
	
	public static boolean unixToStructured(int[] intOut, Object nullableTimeZone, double unixTime) {
		java.util.TimeZone tz = getTimeZone(nullableTimeZone);
		int unixInt = (int) unixTime;
		int micros = (int)(1000000 * (unixTime - unixInt));
		int millis = micros / 1000;
		java.util.Date t = new java.util.Date(1000L * (long) unixInt);
		java.time.ZonedDateTime zdt = t.toInstant().atZone(tz.toZoneId());
		java.time.LocalDate ld = zdt.toLocalDate();
		java.time.LocalTime lt = zdt.toLocalTime();
		
		intOut[0] = ld.getYear();
		intOut[1] = ld.getMonthValue();
		intOut[2] = ld.getDayOfMonth();
		intOut[3] = lt.getHour();
		intOut[4] = lt.getMinute();
		intOut[5] = lt.getSecond();
		intOut[6] = millis;
		intOut[7] = micros;
		intOut[8] = 1 + ld.getDayOfWeek().getValue();
		
		return true;
	}
	
	private static String pad(int value, int size) {
		String output = value + "";
		while (output.length() < size) {
			output = "0" + output;
		}
		return output;
	}
	
	private static java.time.format.DateTimeFormatter parseFormat = java.time.format.DateTimeFormatter.ofPattern("y M d H m s VV");
	public static void parseDate(int[] intOut, Object nullableTimeZone, int year, int month, int day, int hour, int minute, int microseconds) {
		intOut[0] = 1;
		int seconds = (int)(microseconds / 1000000);
		int micros = microseconds - seconds * 1000000;
		intOut[2] = micros;
		boolean isUtc = nullableTimeZone == null;
		java.util.TimeZone tz = getTimeZone(nullableTimeZone);
		String tzId = tz.toZoneId().getId();
		java.time.ZonedDateTime zdt;
		if (isUtc) {
			String tsIso = year + " " + pad(month, 2) + " " + pad(day, 2) + " " + pad(hour, 2) + " " + pad(minute, 2) + ":" + pad(day, 2) + "Z";
			zdt = java.time.ZonedDateTime.parse(tsIso);
		} else {
			String tsParse = year + " " + month + " " + day + " " + hour + " " + minute + " " + seconds + " " + tz.toZoneId().getId();
			zdt = java.time.ZonedDateTime.parse(tsParse, parseFormat);
		}
		intOut[1] = (int) zdt.toEpochSecond();
	}
	
	public static Object[] initializeTimeZoneList() {
		throw new RuntimeException("Not implemented.");
	}
	
	public static boolean isDstOccurringAt(Object nullableTimeZone, int unixTime) {
		throw new RuntimeException("Not implemented.");
	}
}
